storage/replicate_queue.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Ben Darnell
+
+package storage
+
+import (
+	"time"
+
+	"github.com/cockroachdb/cockroach/gossip"
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+const (
+	// replicateQueueMaxSize is the max size of the split queue.
+	replicateQueueMaxSize = 100
+
+	// replicateQueueTimerDuration is the duration between replication of queued ranges.
+	replicateQueueTimerDuration = 0 * time.Second // zero duration to process replication greedily
+)
+
+// replicateQueue manages a queue of ranges to have their replicas
+// change to match the zone config.
+type replicateQueue struct {
+	*baseQueue
+	gossip    *gossip.Gossip
+	allocator *allocator
+	disabled  bool
+}
+
+// newReplicateQueue returns a new instance of replicateQueue.
+func newReplicateQueue(gossip *gossip.Gossip, allocator *allocator) *replicateQueue {
+	rq := &replicateQueue{
+		gossip:    gossip,
+		allocator: allocator,
+	}
+	rq.baseQueue = newBaseQueue("replicate", rq, replicateQueueMaxSize)
+	return rq
+}
+
+func (rq *replicateQueue) shouldQueue(now proto.Timestamp, rng *Range) (
+	shouldQ bool, priority float64) {
+	// Only queue for replication if this replica is the leader.
+	if !rng.IsLeader() {
+		return
+	}
+
+	// If the range spans multiple zones, ignore it until the split queue has processed it.
```

Comment:
What's your thinking here? This doesn't really seem like a reason to ignore a low replication count. They won't conflict due to our mutex in the range and it seems like there's not any cost no matter which order which we do the split / replicate tasks in. Further, while this might determine a range in need of splitting, there's no symmetric check for a merge which might need to occur.

My feeling is we should just remove this check and proceed optimistically.